spotbugs/src/main/java/edu/umd/cs/findbugs/detect/PublicAttributes.java

```diff
@@ -0,0 +1,210 @@
+package edu.umd.cs.findbugs.detect;
+
+import java.util.Set;
+import java.util.HashSet;
+
+import org.apache.bcel.Const;
+import org.apache.bcel.Repository;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Field;
+import org.apache.bcel.generic.ObjectType;
+import org.apache.bcel.generic.Type;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.Detector;
+import edu.umd.cs.findbugs.ba.XField;
+import edu.umd.cs.findbugs.ba.XMethod;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+public class PublicAttributes
+        extends OpcodeStackDetector
+        implements Detector {
+
+    private final BugReporter bugReporter;
+
+    private Set<XField> writtenFields = new HashSet<XField>();
+
+    public PublicAttributes(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    // Check for each statements which write an own attribute of the class
+    // instance. If the attribute written is public then report a bug.
+    @Override
+    public void sawOpcode(int seen) {
+        // It is normal that classes used as simple data types have a
+        // constructor to make initialization easy.
+        if (isConstructorLikeMethod(getMethodName())) {
+            return;
+        }
+
+        // We do not care for enums and nested classes.
+        if (getThisClass().isEnum() || getClassName().indexOf('$') >= 0) {
+            return;
+        }
+
+        if (seen == Const.PUTFIELD || seen == Const.PUTSTATIC) {
+            XField field = getXFieldOperand();
+            // Do not report the same public field twice.
+            if (field == null || writtenFields.contains(field)) {
+                return;
+            }
+
+            // Only consider attributes of self.
```

Comment:
Do we need this limitation?
Actually, the page describing this rule has no limitation like this.

Reply:
Do you mean that we only report public attributes which are written by the methods of the class? We introduced this limitation to reduce noise of the detector and make this rule more realistic to follow. Completely forbidding non-final public attributes seems overkill and results in many false positives. Our first try (which resulted in a much simpler detector but with more false positives) was that if the class is something like a C struct, thus it has no methods at all maybe except a constructor then we do not check it for public attributes. Even this did not seem enough because we found that some classes may have flags that can be set from outside and influence the behavior of the class. E.g. a data member called `DEBUG` may be set from outside to change logging level but methods of the class itself only read it. However, if the class itself writes an attributes that means that the class is responsible for that attribute and thus it should be a private (or protected member).